src/tools/pkcs11-tool.c

```diff
@@ -2382,18 +2407,38 @@ static void verify_signature(CK_SLOT_ID slot, CK_SESSION_HANDLE session,
 	fprintf(stderr, "Using signature algorithm %s\n", p11_mechanism_to_name(opt_mechanism));
 	memset(&mech, 0, sizeof(mech));
 	mech.mechanism = opt_mechanism;
-	hashlen = parse_pss_params(session, key, &mech, &pss_params);
-	if (hashlen && opt_salt_len_given) {
-		if (opt_salt_len == -2) {
-			/* openssl allow us to set sLen to -2 for autodetecting salt length
-			 * here maximal CK_ULONG value is used to pass this special code
-			 * to openssl. For non OpenSC PKCS#11 module this is minimal limitation
-			 * because there is no need to use extra long salt length.
-			 */
-			pss_params.sLen = ((CK_ULONG) 1 ) << (sizeof(CK_ULONG) * CHAR_BIT -1);
-			fprintf(stderr, "Warning, requesting salt length recovery from signature (supported only in in opensc pkcs11 module).\n");
+
+	if (opt_mechanism == CKM_RSA_PKCS_PSS) {
+		hashlen = parse_pss_params(session, key, &mech, &pss_params);
+		if (hashlen && opt_salt_len_given) {
+			if (opt_salt_len == -2) {
+				/* openssl allow us to set sLen to -2 for autodetecting salt length
+				* here maximal CK_ULONG value is used to pass this special code
+				* to openssl. For non OpenSC PKCS#11 module this is minimal limitation
+				* because there is no need to use extra long salt length.
+				*/
+				pss_params.sLen = ((CK_ULONG) 1 ) << (sizeof(CK_ULONG) * CHAR_BIT -1);
+				fprintf(stderr, "Warning, requesting salt length recovery from signature (supported only in in opensc pkcs11 module).\n");
+			}
+		}
+	}
+
+	/* support pure EdDSA only */
+	if (opt_mechanism == CKM_EDDSA) {
+		CK_ULONG pubkeylen = 0;
+
+		/* determine if we have ed25519 or ed448 keys */
+		if (getEC_POINT(session, key, &pubkeylen) == NULL) {
+			util_fatal("Key has no CKA_EC_POINT attribute");
+		}
+
+		/* Ed448, need the params defined but default to false */
+		if (pubkeylen == 57) {
```

Comment:
I do not like the key type detection by the key length. We should really parse the EC_PARAMS to decide on the key type (used also in the first part).

Reply:
Are you good with matching the EC_PARAMS using the ec_curve_info and then checking the size field which would be either 255 or 448 depending on key type?